src/views/ImportCredentials/ImportCredentials.tsx

```diff
@@ -0,0 +1,135 @@
+import { useIdentities } from '../../utilities/identities/identities';
+import { sameFullDid } from '../../utilities/did/did';
+
+import { ImportCredentialsForm } from './ImportCredentialsForm';
+import { ImportCredentialsResults } from './ImportCredentialsResults';
+import { FailedImport, Import, SuccessfulImport } from './types';
+
+export function ImportCredentials(): JSX.Element | null {
+  const [processing, setProcessing] = useState(false);
+  const [pending, setPending] = useState<Import[]>([]);
+  const [failedImports, setFailedImports] = useState<FailedImport[]>([]);
+  const [successfulImports, setSuccessfulImports] = useState<
+    SuccessfulImport[]
+  >([]);
+
+  const identities = useIdentities().data;
+  const identitiesList =
+    identities && sortBy(Object.values(identities), 'index');
+
+  const grouped = useMemo(() => {
+    if (!identitiesList) {
+      return null;
+    }
+    return identitiesList
+      .map((identity) => ({
+        identity,
+        imports: filter(successfulImports, {
+          identityAddress: identity.address,
+        }),
+      }))
+      .filter(({ imports }) => imports.length > 0);
+  }, [identitiesList, successfulImports]);
+
+  const handleFiles = useCallback(
+    (files: FileList) => {
+      const filesList = [...files];
+
+      if (!identitiesList) {
+        return;
+      }
+
+      const pending = filesList.map((file) => ({ fileName: file.name }));
+      setPending(pending);
+      setProcessing(true);
+
+      filesList.forEach(async (file) => {
+        const fileName = file.name;
+        try {
+          const text = await file.text();
+          const credential = JSON.parse(text) as Credential;
+
+          const {
+            request,
+            name = fileName,
+            status = 'pending',
+            cTypeTitle,
+            attester,
+          } = credential;
+
+          if (
+            !cTypeTitle ||
+            !attester ||
+            !RequestForAttestation.verifyData(request)
+          ) {
+            throw new Error('invalid');
+          }
+
+          const knownIdentity = identitiesList.find(({ did }) =>
+            sameFullDid(did, request.claim.owner),
+          );
+          if (!knownIdentity) {
+            throw new Error('orphaned');
+          }
+
+          await saveCredential({
+            request,
+            name,
+            status,
+            cTypeTitle,
+            attester,
+            isDownloaded: true,
+          });
+
+          setSuccessfulImports((successfulImports) => [
+            ...successfulImports,
+            { fileName, identityAddress: knownIdentity.address },
+          ]);
+        } catch (exception) {
+          const error = exceptionToError(exception);
+          setFailedImports((failedImports) => [
+            ...failedImports,
+            { fileName, error: error.message },
+          ]);
+        } finally {
+          setPending((pending) => remove(pending, { fileName }));
```

Comment:
`remove` returns an array of the removed elements, which I don't think is what we want here.

This worked for me, but maybe there's a better method with lodash?

```
setPending(
            pending.filter(({ fileName }) => {
              fileName !== file.name;
            }),
          );
```

Reply:
Yep, there’s `filter(pending, { fileName })`